File: Year II/First Semester/MAP/Seminar/seminar1/seminar1/src/decorator/AbstractTaskRunner.java
package decorator;

import model.Task;

public abstract class AbstractTaskRunner implements TaskRunner{

    // Definiti clasa abstractă AbstractTaskRunner [Decorator Pattern]
    // care implementează interfața TaskRunner si care conține
    // ca și atribut privat o referință la un obiect de tipul Task Runner,
    // referința primită ca parametrul prin intermediul constructorului.
    private TaskRunner taskRunner;

    public AbstractTaskRunner(TaskRunner runner){
        taskRunner = runner;
    }

    @Override
    public void executeOneTask() {
        taskRunner.executeOneTask();
    }

    @Override
    public void executeAll() {
        taskRunner.executeAll();
    }

    @Override
    public void addTask(Task t) {
        taskRunner.addTask(t);
    }

    @Override
    public boolean hasTask() {
        return taskRunner.hasTask();
    }
}

File: Year II/First Semester/MAP/Laborator/Laborator9/ConsoleApp1/ConsoleApp1/Decorator/AbstractTaskRunner.cs
using Task = ConsoleApp1.Model.Task;

namespace ConsoleApp1.Decorator;

public abstract class AbstractTaskRunner: TaskRunner
{
    protected readonly TaskRunner TaskRunner;

    public AbstractTaskRunner(TaskRunner taskRunner)
    {
        TaskRunner = taskRunner;
    }

    public void AddTask(Task task)
    {
        TaskRunner.AddTask(task);
    }

    public bool HasTasks()
    {
        return TaskRunner.HasTasks();
    }

    public abstract void ExecuteOneTask();

    public void ExecuteAll()
    {
        while (HasTasks())
        {
            ExecuteOneTask();
        }
    }
}